flow/layers/picture_layer.cc

```diff
@@ -17,6 +18,68 @@ PictureLayer::PictureLayer(const SkPoint& offset,
       is_complex_(is_complex),
       will_change_(will_change) {}
 
+void PictureLayer::Diff(DiffContext* context, const Layer* old_layer) {
+  DiffContext::AutoSubtreeRestore subtree(context);
+  if (!context->IsSubtreeDirty()) {
+    assert(old_layer);
+    auto prev = old_layer->as_picture_layer();
+    if (!Compare(this, prev)) {
+      context->MarkSubtreeDirty(prev->paint_region());
+    }
+  }
+  context->PushTransform(SkMatrix::Translate(offset_.x(), offset_.y()));
+  context->AddPaintRegion(picture()->cullRect());
+  set_paint_region(context->CurrentSubtreeRegion());
+}
+
+bool PictureLayer::Compare(const PictureLayer* l1, const PictureLayer* l2) {
+  const auto& pic1 = l1->picture_.get();
+  const auto& pic2 = l2->picture_.get();
+  if (pic1.get() == pic2.get()) {
+    return true;
+  }
+  auto op_cnt_1 = pic1->approximateOpCount();
+  auto op_cnt_2 = pic2->approximateOpCount();
+  if (op_cnt_1 != op_cnt_2 || pic1->cullRect() != pic2->cullRect()) {
+    return false;
+  }
+
+  if (op_cnt_1 > 10) {
+    return false;
+  }
+
+  // TODO(knopp) we don't actually need the data; this could be done without
+  // allocations by implementing stream that calculates SHA hash and
+  // comparing those hashes
+  auto d1 = l1->SerializedPicture();
+  auto d2 = l2->SerializedPicture();
+  auto res = d1->equals(d2.get());
+  return res;
+}
```

Comment:
We might want to keep per-frame statistics for:

- number of pictures new and not matching previous frame
- number of pictures recreated, but identical to previous
- number of pictures reused by reference

Also, for future work on the framework we might want to expose these comparisons via the native->Dart interface on Picture so that we can do some debugging and analysis at the framework level on when we are creating duplicate pictures unnecessarily. But if our statistics show that most apps only have either brand new, or reused pictures and rarely have a picture that was recreated, but identical then we may not need to do that kind of debugging.

Reply:
Indeed. Especially knowing the number of pictures that are different instance but same contents might be useful. I don't think there is any infrastructure in place to gather statistics like this, is there?

Reply:
I suppose we can just trace it to timeline for now.

Reply:
We can inject statistics into the Trace logs and do so to indicate the number of picture layers cached, for instance.

If you had something here to record the statistic then at the top after the Diff of the tree is done you could inject a Trace event with the 3 numbers.

Look at this code in rasterizer.cc for an example:

```
    fml::tracing::TraceEventAsyncComplete(
        "flutter",                    // category
        "SceneDisplayLag",            // name
```

Or, `RasterCache::TraceStatsToTimeline` for another example.

Reply:
Our replies crossed as we were writing. I agree, this seems like a Timeline thing. I think one summary event per frame should be enough.

Reply:
[This commit](https://github.com/flutter/engine/pull/21498/commits/399c6bffefa93c1e1609a4c1f2cf94708b774df9) added the following statistics:
```cpp
    // Picture replaced by different picture
    void AddNewPicture();

    // Picture that would require deep comparison but was considered too complex
    // to serialize and thus was treated as new picture
    void AddPictureTooComplexToCompare();

    // Picture that has identical instance between frames
    void AddSameInstancePicture();

    // Pictures that had to be serialized to compare for equality
    void AddDeepComparePicture();

    // Pictures that had to be serialized to compare (different instances),
    // but were equal
    void AddDifferentInstanceButEqualPicture();
```

Reply:
I've been informed that the SkPicture serializer may not be available in all build modes. Have you tested this with debug/profile/release builds?

Reply:
I have, with the first version of diff (but it still used serialization). I'll retest it.

Reply:
I tried release build on iOS and the serializer works as expected.